### PR TITLE
Fixed Shadergraph Unlit Custom Shadows

### PR DESCRIPTION
**Pull request by lukaschod**
### Purpose of this PR
Fixed unlit shadergraph custom shadows to not print error about missing shadowCoord (https://fogbugz.unity3d.com/f/cases/1175274/).
Now BuilVarryings will include shadowCoord if it is actually used in target, also making consistant with other properties. Also shadowCoord is now conditional in lit shadergraph like in lit shader.

---
### Testing status
Added new test 146_Unlit_Custom_Shadows to prevent regression.
image
Basically very simple custom shadows in unlit shadergraph

### CODE REVIEW

**Review by phi-lira — APPROVED** [review]
LGTM. Although it's not clear to me how we test screen space path.

**Comment by lukaschod** [comment]
> LGTM. Although it's not clear to me how we test screen space path.

I am not testing screen space path, but just custom shadows implemented with (GetMainLight) on unlit shader.
This way making sure users that try to implement themselves similar behaviour does not get regressed.

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Scenes/146_Unlit_Custom_Shadows/CustomLighting.hlsl [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+   Color = mainLight.color;
+   DistanceAtten = mainLight.distanceAttenuation;
+   ShadowAtten = mainLight.shadowAttenuation;
+   #endif
+}
+
+void SampleSH_half(half3 normalWS, out half3 Ambient)
+{
+    // LPPV is not supported in Ligthweight Pipeline
+    real4 SHCoefficients[7];
```
Comment (nigeljw-unity) [inline TestProjects/UniversalGraphicsTest_Lighting/Assets/Scenes/146_Unlit_Custom_Shadows/CustomLighting.hlsl:61]: Can you move these assignments into an array initializer list so that it is just one line?

Comment (nigeljw-unity) [reply TestProjects/UniversalGraphicsTest_Lighting/Assets/Scenes/146_Unlit_Custom_Shadows/CustomLighting.hlsl:61]: It is brutal how in the function, it just splits them out as well. Having too many parameters can impact compilation and fitting things in registers, but I can't see this as being better. 
float3 SampleSH9(float4 SHCoefficients[7], float3 N)
{
    float4 shAr = SHCoefficients[0];
    float4 shAg = SHCoefficients[1];
    float4 shAb = SHCoefficients[2];
    float4 shBr = SHCoefficients[3];
    float4 shBg = SHCoefficients[4];
    float4 shBb = SHCoefficients[5];
    float4 shCr = SHCoefficients[6];

Comment (nigeljw-unity) [reply TestProjects/UniversalGraphicsTest_Lighting/Assets/Scenes/146_Unlit_Custom_Shadows/CustomLighting.hlsl:61]: Do you know if there is a reason it is done this way? Or can we just modify the helper function to remove the temporaries?

File: com.unity.render-pipelines.universal/Editor/ShaderGraph/Includes/Varyings.hlsl [hunk 2]
```diff
@@ -119,7 +119,7 @@ Varyings BuildVaryings(Attributes input)
     output.fogFactorAndVertexLight = half4(fogFactor, vertexLight);
 #endif
 
-#if defined(REQUIRES_VERTEX_SHADOW_COORD_INTERPOLATOR)
+#if defined(VARYINGS_NEED_SHADOW_COORD) && defined(REQUIRES_VERTEX_SHADOW_COORD_INTERPOLATOR)
```
[the remark this answers is not in the document]
Comment (lukaschod) [reply com.unity.render-pipelines.universal/Editor/ShaderGraph/Includes/Varyings.hlsl:122]: Sadly not, but maybe there is better solution.

VARYINGS_NEED_SHADOW_COORD is currently defined by shadergraph
REQUIRES_VERTEX_SHADOW_COORD_INTERPOLATOR is currently defined by our shadows include

In shadergraph the logic requires these both conditions:
- Property is available in target.
- Current pass uses it.

Maybe we need better naming :/, but that again can cause regression

File: TestProjects/UniversalGraphicsTest_Lighting/Assets/Scenes/146_Unlit_Custom_Shadows/CustomLighting.hlsl (hunk 1 above)
Comment (lukaschod) [reply TestProjects/UniversalGraphicsTest_Lighting/Assets/Scenes/146_Unlit_Custom_Shadows/CustomLighting.hlsl:61]: IDK this is from our URP as these a test function. I think GPU will eleminate these temporaries any way

**Review by Verasl — APPROVED** [review]
It works!! nice. LGTM, and tests 👍

**Comment by lukaschod** [comment]
[Yamato] ShaderGraph_Stereo on Win__Standalone_cache_mono_Linear on version trunk — Failed
[09:09:21.576 INF] System.AggregateException: One or more errors occurred. (The system cannot find the file specified.)
 ---> System.ComponentModel.Win32Exception (2): The system cannot find the file specified.
   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)
   at System.Diagnostics.Process.Start()

[Yamato] URP_Foundation on Win_DX11_Standalone_XR_cache_mono_Linear on version trunk
[09:37:56.993 INF] System.AggregateException: One or more errors occurred. (The system cannot find the file specified.)
 ---> System.ComponentModel.Win32Exception (2): The system cannot find the file specified.
   at System.Diagnostics.Process.StartWithCreateProcess(ProcessStartInfo startInfo)

[Yamato] URP_Lighting on OSX_Metal_playmode_cache_mono_Linear on version trunk 
[08:35:04.886 INF] Starting process on subprocess runner
[08:35:04.990 INF] Warning: Permanently added '172.28.224.189' (ECDSA) to the list of known hosts.
[08:38:10.632 INF] packet_write_poll: Connection to 172.28.224.189 port 22: Operation not permitted
[08:38:10.634 INF] Process exited

https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/universal%252Fsg-unlit-shadow-coord-fix/.yamato%252Funiversal_hybrid-win-dx11.yml%2523Universal_Hybrid_Win_DX11_playmode_mono_Linear_trunk/5532979/job
[Yamato] Universal_Hybrid on OSX_Metal_playmode_mono_Linear on version trunk
https://yamato.cds.internal.unity3d.com/jobs/902-Graphics/tree/master/.yamato%252Funiversal_hybrid-osx-metal.yml%2523Universal_Hybrid_OSX_Metal_playmode_mono_Linear_trunk/5553799/job
[08:38:10.639 INF] lost connection
   at System.Diagnostics.Process.Start()
